nixos/modules/services/networking/zerotierone.nix

```diff
@@ -39,20 +39,47 @@ in
     systemd.services.zerotierone = {
       description = "ZeroTierOne";
       path = [ cfg.package ];
+      confinement = {
+        enable = true;
+        packages = [ pkgs.coreutils ];
+        mode = "chroot-only";
+      };
       bindsTo = [ "network-online.target" ];
       after = [ "network-online.target" ];
       wantedBy = [ "multi-user.target" ];
       preStart = ''
         mkdir -p /var/lib/zerotier-one/networks.d
-        chmod 700 /var/lib/zerotier-one
-        chown -R root:root /var/lib/zerotier-one
       '' + (concatMapStrings (netId: ''
         touch "/var/lib/zerotier-one/networks.d/${netId}.conf"
       '') cfg.joinNetworks);
+
       serviceConfig = {
         ExecStart = "${cfg.package}/bin/zerotier-one -p${toString cfg.port}";
         Restart = "always";
         KillMode = "process";
+        StateDirectory = "zerotier-one";
+        StateDirectoryMode = "0700";
+
+        CapabilityBoundingSet = "CAP_NET_ADMIN CAP_NET_RAW";
+        LockPersonality = true;
+        NoNewPrivileges = true;
+        PrivateTmp = true;
+        ProtectControlGroups = true;
+        ProtectKernelModules = true;
+        ProtectKernelTunables = true;
```

Comment:
Please note that this implies `MountAPIVFS`, so you could just go with the default confinement mode of `full-apivfs` and remove these options.